### PR TITLE
GL E-Ink Permanent Config form validation

### PR DESCRIPTION
**Pull request by PaulJKim**
**Asana task**: [[Perm config] Form validation](https://app.asana.com/0/1185117109217413/1205918501622291/f)

Description
This PR implements form validation for GL E-ink permanent config.

### CODE REVIEW

**Comment by cmaddox5** [comment]
FYI, I resolved merge conflicts before review to help with testing the changes.

File: lib/screenplay/config/config.ex [hunk 1]
```diff
@@ -87,6 +92,25 @@ defmodule Screenplay.Config.PermanentConfig do
     end
   end
 
+  defp check_for_duplicate_screen_ids(deserialized, places_and_screens) do
+    %PendingConfig{screens: existing_screens} = PendingConfig.from_json(deserialized)
+
+    all_screen_ids =
```
Comment (cmaddox5) [inline lib/screenplay/config/config.ex:98]: We should include the live screens in this list as well. There's technically no issue with a pending and live screen sharing an ID at this point, but it will be an issue when the pending screen is published.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx [hunk 2]
```diff
@@ -32,6 +49,86 @@ const GlEinkWorkflow: ComponentType<WorkflowProps> = ({
...
+  };
+
+  const validateRequiredFields = (placesAndScreens: PlaceIdsAndNewScreens) => {
+    const fieldsWithErrors = new Set<string>();
+    for (const [place_id, screens] of Object.entries(placesAndScreens)) {
+      const fieldsByScreen = screens["new_pending_screens"]?.map((screen) => {
+        const presentFields = [];
+
+        // Validate that screen id is in correct format
+        if (screen.new_id?.match(/^EIG-\d+$/)) {
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:99]: Should we consider the ID field invalid if it is still set to the default of `EIG-`? That is technically a valid ID, but more than likely always be a mistake if it is ever submitted without numbers after the hyphen.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx [hunk 3]
```diff
@@ -66,24 +164,88 @@ const GlEinkWorkflow: ComponentType<WorkflowProps> = ({
...
+                  setShowValidationAlert(true);
+                  dispatch({
+                    type: "SET_VALIDATION_ERRORS",
+                    validationErrors: validationErrors,
+                  });
+                })
+                .catch((error) => {
+                  console.log(error);
+                });
+            }
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:210]: Can this be extracted into its own function? The nested anonymous functions are getting a little hard to follow.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/ConfigureScreensPage.tsx [hunk 4]
```diff
@@ -327,6 +369,7 @@ const ConfigureScreenRow: ComponentType<ConfigureScreenRowProps> = ({
           }}
           placeholder="EIG-"
         />
+        {screenIdError()}
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/ConfigureScreensPage.tsx:372]: This column looks like it needs a little more width so the error text doesn't wrap:
Screenshot 2024-03-04 at 9 46 15 AM

**Review by cmaddox5 — CHANGES_REQUESTED** [review]
Digging the approach here. Visually (outside of one wrapping issue) things look great. Have a few comments on some of the validation logic.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx (hunk 2 above)
Comment (PaulJKim) [reply assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:99]: Yeah I wondered about that as well. The designs technically do suggest that `EIG-` should be treated as an invalid ID though so I just went with that https://www.figma.com/file/FKWRVrSdoMjZOLorhuykvc/Permanent-Configuration?type=design&node-id=1622-23109&mode=design&t=eKbCkI2xTlb8X1UU-4 let me know if you think we should push back on that though

Comment (cmaddox5) [reply assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:99]: Ahh, I misunderstood your regex here. I thought it was looking for at least `EIG-` but it actually is doing what I was thinking in my comment. My bad, disregard.

**Review by cmaddox5 — APPROVED** [review]
Looking good!

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/ConfigureScreensPage.tsx [hunk 5]
```diff
@@ -65,12 +71,33 @@ const ConfigureScreensWorkflowPage: ComponentType<ConfigureScreensWorkflowPagePr
...
+      useConfigValidationContext();
+    const dispatch = useConfigValidationDispatchContext();
+
     useEffect(() => {
       if (selectedPlaces.length) {
         fetchExistingScreens(
           "gl_eink_v2",
           selectedPlaces.map((place) => place.id)
         ).then(({ places_and_screens, version_id }) => {
+          for (const place_id in places_and_screens) {
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/ConfigureScreensPage.tsx:84]: Can this block be refactored into a named function?

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx [hunk 6]
```diff
@@ -66,24 +208,72 @@ const GlEinkWorkflow: ComponentType<WorkflowProps> = ({
         navigate(-1);
       };
       onBack = () => {
+        Object.keys(placesAndScreensToUpdate).forEach((placeId) => {
+          newScreenValidationErrors[placeId] = [];
+        });
+        Object.keys(placesAndScreensToUpdate).forEach((placeId) => {
+          pendingScreenValidationErrors[placeId] = [];
+        });
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:216]: Can these two be combined?

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx [hunk 7]
```diff
@@ -66,24 +208,72 @@ const GlEinkWorkflow: ComponentType<WorkflowProps> = ({
...
+            newScreenValidationErrors,
+            pendingScreenValidationErrors,
+          });
+        }
       };
       layout = (
-        <ConfigureScreensWorkflowPage
-          selectedPlaces={places.filter((place) =>
-            selectedPlaces.has(place.id)
+        <div>
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:251]: Does this `div` help with styling at all? If not, it can be converted to a fragment.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx [hunk 8]
```diff
@@ -66,24 +208,72 @@ const GlEinkWorkflow: ComponentType<WorkflowProps> = ({
...
+            selectedPlaces={places.filter((place) =>
+              selectedPlaces.has(place.id)
+            )}
+            setPlacesAndScreensToUpdate={setPlacesAndScreensToUpdate}
+            handleRemoveLocation={handleRemoveLocation}
+            setConfigVersion={setConfigVersion}
+          />
+          {validationErrorMessage !== "" && (
+            <div className="config-validation-alert-container">
+              <Alert
```
Comment (cmaddox5) [inline assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:262]: I noticed the close button hover effect looks a bit off because its positioning is towards the top instead of being centered vertically. Should the hover effect bg color be transparent to help with this, or is this by design?

**Review by cmaddox5 — APPROVED** [review]
Looks good! Just have some questions and suggestions, but nothing blocking so will approve now.

File: assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx (hunk 8 above)
Comment (PaulJKim) [reply assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/GlEinkWorkflow.tsx:262]: Not by design, thank you for that suggestion! Definitely looks better now
